### DISCUSSION TITLE
Custom Webpack Config -- Allow Global CSS

### DISCUSSION THREAD

**Discussion by brianabaker** [vercel/next.js#28666] [Help]
Hello! 

I'm on Nextjs v `10.0.0`. We're currently using `@zeit/next-sass` and I'm trying to migrate off of it. I understand there's hard rule around not allowing css imports outside of the `app` page now, but that's not going to work for me. 

I have a **lot** of global css imports in my application, and I need to allow both global css imports and css modules at the same time while converting everything to css modules. I've narrowed the error down to [this line in this file](https://github.com/TrySound/next.js/blob/9821e149296d31e978c897aea16dee44588c768c/packages/next/build/webpack/config/blocks/css/index.ts#L268), but I can't figure out how to override it. 

I've found a bunch of different discussions and PRs, but none have worked for my use case thus far. 

I am not asking about enabling it for node modules at all. I want to use global css in regular js files. 

Can I modify the webpack config to do this? I've tried a few hacks I found online but none of have worked so far. 

Example of what I have in my `next.config.js` trying to do this:
```
    const serverCSSTest = [regexCssGlobal, regexSassGlobal];

    config.module.rules.forEach((rule) => {
      if (typeof rule.oneOf !== "object") {
        return rule;
      }

      rule.oneOf.forEach((oneOf) => {
        if (oneOf.test.toString() === serverCSSTest.toString()) {
          if (
            oneOf.use &&
            oneOf.use.options &&
            oneOf.use.options.reason ===
              "Global CSS \u001b[1mcannot\u001b[22m be imported from files other than your \u001b[1mCustom <App>\u001b[22m. Please move all global CSS imports to \u001b[36mpages/_app/index.js\u001b[39m. Or convert the import to Component-Level CSS (CSS Modules).\n" +
                "Read more: https://err.sh/next.js/css-global"
          ) {
            // What do I need to do here to stop this
          }
        }
      });
    });

    return config;
  },
```

References:
[Merged PR allowing css in node modules](https://github.com/vercel/next.js/pull/16993/files#diff-f14cc6d633f9cc94b52658a427dba0e821f47b87b1e33a1ddab4c76e34adfc40L162)
[Request to expose default loaders in webpack config](https://github.com/vercel/next.js/issues/24140)
[RFC for allowing global css for node modules](https://github.com/vercel/next.js/discussions/27953)
[A hack I found on a package PR](https://github.com/nrwl/nx/issues/3803)

**Comment by brianabaker** [comment]
I got it working by editing the source code at `next/dist/build/webpack/config/blocks/css/index.js`, but would like to find another way.

**Comment by alebelcor** [comment]
I think I'm in a somewhat similar situation to yours. I want to re-enable support for CSS Modules.

I looked at [this plugin's source](https://github.com/bem/next-global-css) trying to adapt it for CSS Modules but couldn't make it work. Maybe it'll work for your global CSS use case 🤞 

In the end I was able to do it by looking at Next.js' source to see what was happening, what was taken out, and put it back in. I have it isolated into a Next.js "plugin":

```js
const postCssConfigLoader = require('postcss-load-config');

const { getCssModuleLoader } = require('next/dist/build/webpack/config/blocks/css/loaders/modules');

// may be different on newer versions
const {
  default: MiniCssExtractPlugin,
} = require('next/dist/build/webpack/plugins/mini-css-extract-plugin/src/index');

const { CssMinimizerPlugin } = require('next/dist/build/webpack/plugins/css-minimizer-plugin');

module.exports = (nextConfig = {}) => {
  // empty array if you don't need PostCSS
  const { plugins: postCssPlugins } = postCssConfigLoader.sync({});

  return Object.assign({}, nextConfig, {
    webpack: (config, options) => {
      const isServer = options.isServer;
      const isClient = !isServer;
      const isDevelopment = options.dev;
      const isProduction = !isDevelopment;
      const assetPrefix = options.config.assetPrefix;

      // restore CSS Modules
      config.module.rules.push(
        {
          sideEffects: false,
          test: /\.module\.css$/,
          use: getCssModuleLoader(
            {
              isDevelopment,
              isClient,
              isServer,
              assetPrefix,
            },
            postCssPlugins
          ),
        }
      );

      // restore minification
      config.optimization.minimizer.push((compiler) => {
        new CssMinimizerPlugin({
          postcssOptions: {
            map: {
              inline: false,
              annotation: false,
            },
          },
        }).apply(compiler);
      });

      if (isClient && isProduction) {
        // restore extraction
        config.plugins.push(
          new MiniCssExtractPlugin({
            filename: 'static/css/[contenthash].css',
            chunkFilename: 'static/css/[contenthash].css',
            ignoreOrder: true,
          })
        );
      }

      if (typeof nextConfig.webpack === 'function') {
        return nextConfig.webpack(config, options);
      }

      return config;
    },
  });
};
```

I'm sure you could do the same with global CSS (i.e. [`getGlobalCssLoader`](https://github.com/vercel/next.js/blob/canary/packages/next/build/webpack/config/blocks/css/loaders/global.ts#L8)) and/or with Sass.

Good luck.
